source/extensions/filters/common/local_ratelimit/local_ratelimit_impl.cc

```diff
@@ -32,28 +33,52 @@ LocalRateLimiterImpl::~LocalRateLimiterImpl() {
 }
 
 void LocalRateLimiterImpl::onFillTimer() {
+  onFillTimerHelper(tokens_, max_tokens_, tokens_per_fill_);
+  onFillTimerDescriptorHelper();
+  fill_timer_->enableTimer(fill_interval_);
+}
+
+void LocalRateLimiterImpl::onFillTimerHelper(const Token& tokens, const uint32_t max_tokens,
+                                             const uint32_t tokens_per_fill) {
   // Relaxed consistency is used for all operations because we don't care about ordering, just the
   // final atomic correctness.
-  uint32_t expected_tokens = tokens_.load(std::memory_order_relaxed);
+  uint32_t expected_tokens = tokens.tokens.load(std::memory_order_relaxed);
   uint32_t new_tokens_value;
   do {
     // expected_tokens is either initialized above or reloaded during the CAS failure below.
-    new_tokens_value = std::min(max_tokens_, expected_tokens + tokens_per_fill_);
+    new_tokens_value = std::min(max_tokens, expected_tokens + tokens_per_fill);
 
     // Testing hook.
     synchronizer_.syncPoint("on_fill_timer_pre_cas");
 
     // Loop while the weak CAS fails trying to update the tokens value.
-  } while (
-      !tokens_.compare_exchange_weak(expected_tokens, new_tokens_value, std::memory_order_relaxed));
+  } while (!tokens.tokens.compare_exchange_weak(expected_tokens, new_tokens_value,
+                                                std::memory_order_relaxed));
+}
 
-  fill_timer_->enableTimer(fill_interval_);
+void LocalRateLimiterImpl::onFillTimerDescriptorHelper() {
+  auto cur = time_source_.monotonicTime();
+  for (const auto& descriptor : last_refill_per_descriptor_) {
+    if (std::chrono::duration_cast<std::chrono::milliseconds>(cur - descriptor.second) >=
+        descriptor.first.token_bucket_.fill_interval_) {
+      auto it = tokens_per_descriptor_.find(descriptor.first);
+      if (it == tokens_per_descriptor_.end()) {
+        continue;
+      }
```

Comment:
How can this happen? I think the code would be simpler if you had a single map with descriptor as key and a struct with token and last fill?

Reply:
Fixed